File: rust/src/bin/2-2.rs
use std::io::prelude::*;
use std::fs::File;
use std::collections::HashSet;
use std::iter::FromIterator;


static GRID : [&'static[i32; 5]; 5] = [
    &[0, 0, 1, 0, 0],
    &[0, 2, 3, 4, 0],
    &[5, 6, 7, 8, 9],
    &[0,10,11,12, 0],
    &[0, 0,13, 0, 0],
];

fn main() {
    let allowedarr = [(0,2),(1,1),(1,2),(1,3),(2,0),(2,1),(2,2),(2,3),(2,4),(3,1),(3,2),(3,3),(4,2)];
    let allowed = HashSet::<_>::from_iter(allowedarr.iter());

    let mut f = File::open("../inputs/2.txt").unwrap();
    let mut s = String::new();
    f.read_to_string(&mut s).unwrap();
    let len = s.len();
    s.truncate(len - 1);

    let keys: String = s
        .split("\n")
        .map(|tokenlist|
            tokenlist.chars().fold((2,0), |coord, dir| {
                let newcoord = match dir {
                    'U' => (coord.0 - 1, coord.1),
                    'D' => (coord.0 + 1, coord.1),
                    'R' => (coord.0, coord.1 + 1),
                    'L' => (coord.0, coord.1 - 1),
                    _ => panic!("Malformed input")
                };
                if allowed.contains(&newcoord) { newcoord } else { coord }
            })
        )
        .map(|coord| GRID[coord.0 as usize][coord.1 as usize])
        .map(|e| format!("{:x}", e))
        .collect::<Vec<String>>()
        .join("")
        .to_uppercase();

    println!("{}", keys);
}

File: ruby/2-2.rb
@input = File.open('./inputs/2.txt', "r").readlines.each {|l| l.chomp!}.collect! {|e| e.split('')}
@keypadtwo = [[2,3,4], [6,7,8], ["A","B","C"]]
def two
    x = 0
    y = 1
    acc = []

    uplock = false
    downlock = false
    rightlock = false
    leftlock = true
    @input.each do |elem|
        elem.each do |e|
            case e
            when 'U'
                next if (leftlock || rightlock || uplock)
                if downlock
                    downlock = false
                    next
                end
                if y == 0 && x == 1
                    uplock = true
                    next
                end
                y -= 1
            when 'D'
                next if (leftlock || rightlock || downlock)
                if uplock
                    uplock = false
                    next
                end
                if y == 2 && x == 1
                    downlock = true
                    next
                end
                y += 1
            when 'R'
                next if (uplock || downlock || rightlock)
                if leftlock
                    leftlock = false
                    next
                end
                if y == 1 && x == 2
                    rightlock = true
                    next
                end
                x += 1
            when 'L'
                next if (uplock || downlock || leftlock)
                if rightlock
                    rightlock = false
                    next
                end
                if y == 1 && x == 0
                    leftlock = true
                    next
                end
                x -= 1
            end
            x = 0 if x == -1
            x = 2 if x == 3
            y = 0 if y == -1
            y = 2 if y == 3
        end

        if uplock
            acc << 1
        elsif downlock
            acc << 'D'
        elsif rightlock
            acc << 9
        elsif leftlock
            acc << 5
        else
            acc << @keypadtwo[y][x]
        end
    end

    acc.join
end

p two
